FI-3589 Test kit metadata for the platform template

arscan opened the pull request:
# Summary
The purpose of this PR is to reduce effort in maintaining platforms by extracting Test Kit metadata from the Test Kit gems and using that to populate the Test Kit List page ([example](https://inferno.healthit.gov/test-kits/)) and the Test Kit Details page ([example](https://inferno.healthit.gov/test-kits/g10-certification/)).  It effectively pre-generates the test kit markdown files that Jekyll uses to generate these html pages from.  Having just another generation step is preferable to having the test kit list and details page be generated dynamically on each page load, as the infrastructure is already there to publish these as static pages and it involves much fewer changes.

Once this is approved, we'll update the inferno.healthit.gov platform with these changes.

You can try it out by running:

```
bundle exec rake web:serve
```

Note that you cannot make a test session this way; in order to do that you need to run the full application via  `./setup.sh` and `./run.sh`.  These have been updated to create the test kit md files as described below, so it should just work.

The Rakefile is now set up such that `serve` first calls `generate`, which now calls a new rake task called `create_test_kit_pages`, which flushes the `web/_test_kit` directory, checks the `web/_config.yml` file for excluded test kits, then creates the pages for all non-excluded test kits in `web/_test_kit`.  It then continues on to generate the site using Jekyll as normal.

I also had to make a few minor changes to the html template for the test kits, as our hand-written front-matter is a bit more clever because we knew about suites in test kits that happen to all share the same options... its a pain to figure that out in a script so the generated one now just generates options explicitely for each suite in the front matter.  That just required a few small tweaks regarding how options are selected from labels, and how options show as defaults.

arscan committed:
Clear old test kit metadata files.

arscan committed:
Do not save test kit generated metadata

arscan committed:
Create test kit pages from test kit gem metadata.

Jammjammjamm commented on Rakefile:
```diff
@@ -13,11 +14,75 @@ namespace :db do
+    ENV['NO_DB'] = 'true'
+    Inferno::Application.start(:suites)
+
+    FileUtils.rm_rf(Dir.glob("#{TEST_KIT_DIR}/*.md"))
```
Doing this without using `File.join(__dir__, ...)` makes me nervous.

arscan replied:
Yeah, good point.  I was looking at the constant and coudln't think of any way this could possibly mess up which is why I kept my initial quick and dirty solution instead of cleaning it up on the final pass before PR, but when rm_rf involved good to be a bit more careful in case someone messes with that constant in the future.  Fixed, I believe.

arscan replied:
Hmm, I forgot the `Dir.glob`.  Fixed again and I confirmed the delete works by adding in another 'excluded_test_kit' (but didn't commit that).

Jammjammjamm commented on Rakefile:
```diff
@@ -13,11 +14,75 @@ namespace :db do
+    test_kits.each do |test_kit|
+      next if excluded_test_kits.include?(test_kit.id.to_s)
+
+      formatted_id = test_kit.id.to_s.gsub('_', '-').sub(/-?test-kit$/, '')
```
```suggestion
      formatted_id = test_kit.id.to_s.dasherize.delete_suffix('-test-kit')
```
When wouldn't there be a `-` in front of `test-kit`?

Jammjammjamm reviewed (commented):
This looks great.

arscan committed:
Make directory clearing a bit safer.

arscan committed:
Clean up id formatting code

Co-authored-by: Stephen MacVicar <Jammjammjamm@users.noreply.github.com>

arscan committed:
Add back in the Dir.glob for the wildcard when clearing old pages.

Jammjammjamm commented on .gitignore:
```diff
@@ -13,6 +13,7 @@
 .idea
 /dashboard/data/*.db
 /_site/
+/web/_test_kits/
```
Should we be adding an override for `/web/_test_kits/.keep`?

arscan replied:
I'm not sure?  This is consistent with what we did for `/_site/`, I believe.  Though different than what we did for `/data/` as that uses wildcards and subdirs that exclude `.keep` from the .gitignore.

Jammjammjamm commented on Rakefile:
```diff
@@ -13,11 +14,75 @@ namespace :db do
+              {
+                'label' => list_option[:label],
+                'value' => list_option[:value],
+                'default' => list_option[:default]
```
I made a comment which bears on this in this PR: https://github.com/inferno-framework/us-core-test-kit/pull/229

arscan replied:
Updated code to use the method we decided upon in core, so this should be resolved now.

Shaumik-Ashraf commented on README.md:
```diff
@@ -88,6 +83,9 @@ customize it to include the relevant Test Kits and content for your platform.
 * This will allow you to rebuild using Jekyll locally, and serve a copy of the static files
 * You can run `bundle exec rake web:serve` to generate the files and view them at `http://localhost:4000`
```
You need to run `bundle exec rake db:migrate` before web:serve or it won't work

arscan replied:
Fixed!  Also need to run `bundle install` or `bundle` so I added that.  Thanks for a close read of the docs.

Shaumik-Ashraf commented on README.md:
```diff
@@ -88,6 +83,9 @@ customize it to include the relevant Test Kits and content for your platform.
 * This will allow you to rebuild using Jekyll locally, and serve a copy of the static files
 * You can run `bundle exec rake web:serve` to generate the files and view them at `http://localhost:4000`
 * Alternatively, you can run `docker compose run inferno_web bundle exec rake web:generate` to only generate the files
```
This sections a bit confusing because you go back and forth between doing it on Docker and doing it locally. Maybe it should all just be on Docker since this seems very production oriented?

I tried all commands here locally and via Docker and they all worked, so this section is still correct.

arscan replied:
Hmm, i'm not sure why i did the docker version in here.  I removed that.

Shaumik-Ashraf reviewed (commented):
I was able to run everything on Docker with Ruby 3.3.6. I have some comments about the README but otherwise looks good!

Jammjammjamm reviewed (commented):
![Screenshot 2025-02-14 at 10 56 58 AM](https://github.com/user-attachments/assets/ea096dbb-ef9e-4fbd-a064-7bdaf9f4877b)

I don't know why I'm getting this js error when I try to start a session. Things are looking good otherwise.

arscan commented:
> ![Screenshot 2025-02-14 at 10 56 58 AM](https://private-user-images.githubusercontent.com/15969967/413382755-ea096dbb-ef9e-4fbd-a064-7bdaf9f4877b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1NDk5MTYsIm5iZiI6MTczOTU0OTYxNiwicGF0aCI6Ii8xNTk2OTk2Ny80MTMzODI3NTUtZWEwOTZkYmItZWY5ZS00ZmJkLWEwNjQtN2JkYWY5ZjQ4NzdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE0VDE2MTMzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2ODMwYTZiNGQzZTU2YjRiMzE4NTE1ZjVhYzljMmUwMTI3ZDVhYjExZmU0ODg0Y2RhOWNhNDZjM2ZjNjQ1MGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-lplWG7j3j53p0RNbQer0_yRT7ZAyOt_rhZsRU-H--U)
> 
> I don't know why I'm getting this js error when I try to start a session. Things are looking good otherwise.

Oh no!  This was the problem i was having on the ruby version update but that then disappeared when i switched to my Mac (i was skeptical about those 2 files not having the right ruby version would be the cause of that... i think that was a coincidence).  This is working for me now.  Hummm, hard to figure out what is wrong unless we hook up a non-packed version of the react app somehow?

arscan committed:
Address README feedback.

arscan committed:
Point to new us core branch with proper option defaults.

arscan committed:
Update to look at the correct spot for option defaults

arscan committed:
Remove default suite code because that doesnt exist.

arscan commented:
I updated this to point to the branch in us core that has the proper method for option defaults.  You can try it out by confirming that SMART v2 is default for all US Core suites in the UI.

We will need to update the reference to us core to the released version when that is done.

I'm having trouble getting docker to reach out to git to download the branch of US core, so it is hard to run the whole thing until thats done.  Once we put out a new us core release and reference that we can debug the intermittent js error.

~One thing that we might be able to clean up is there seems to be some stickiness with the generated files in the images; I think we might want to look a little closer at how jekyll is run here and ramifications on what gets saved in containers, etc.~ I added a volume to persist changes to _test_kits to the host, which makes it easier to verify changes made by the new create test kit rake task.  This is consistent with what we already do for _site.

arscan committed:
Have test kit pages be in volume so docker users can see changes locally.

arscan committed:
Auto-remove old containers from docker run to avoid orphans.

arscan committed:
Add new validator that we cut over to.

arscan committed:
Fix typo in Dockerfile.

arscan committed:
Move back to published US Core Test Kit now that option defaults implemented.

arscan committed:
Have ID of suite option be string and not Ruby symbol.

arscan commented:
Ok, I believe this is working now as intended.  I made a few changes since last time to docker-related things.

The suite interface now renders correctly for me.  Can you try?  I'm *hoping* it had something to do with a minor fix I made regarding suite options.  But I'm not 100% sure because the issue seemed intermittent regardless.

Something that is worth doing, btw, is running `docker compose down --remove-orphans`; before we didn't have the `--rm` in there so these runs caused a bunch of stopped orphan containers to pile up.  But this should not happen any more.

Shaumik-Ashraf reviewed (approved):
Works for me
